### PR TITLE
extract components with opts and default values

### PR DESCRIPTION
**Pull request by mrhooray**
related to #36, first step of changes that
- expose tunable parameters at initialization & runtime in the future
- unblock benchmark on convergence time with custom configuration
- separate components for easier navigation and future work

### CODE REVIEW

**Comment by jwolski** [comment]
Looks legit, but now we have `lib/swim` dir and `lib/swim.js` which might get confusing. Thoughts about moving `lib/swim.js` to `lib/swim/index.js`?

**Comment by Raynos** [comment]
:+1: more folders.

**Comment by mrhooray** [comment]
@jwolski updated pr with all remaining parts of `lib/swim.js` extracted and expose suspicionTimeout on ringpop creation

File: index.js [hunk 1]
```diff
@@ -106,8 +107,14 @@ function RingPop(options) {
...
     this.memberIterator = new MemberIterator(this);
-    this.gossip = new Gossip(this);
-    this.suspicion = new Suspicion(this);
+    this.gossip = new Gossip({
+        ringpop: this,
+        minProtocolPeriod: options.minProtocolPeriod
+    });
+    this.suspicion = new Suspicion({
+        ringpop: this,
+        minProtocolPeriod: options.suspicionTimeout
```
Comment (jwolski) [inline index.js]: this seems wrong. why does Suspicion need minProtocolPeriod?

Comment (mrhooray) [inline index.js]: good catch, will update

File: test/swim_test.js [hunk 2]
```diff
@@ -181,6 +187,8 @@ test('marks member faulty after suspect period', function t(assert) {
 
     suspicion.start(member);
 
-    suspicion.stopAll();
-    assert.end();
+    setTimeout(function verify() {
```
Comment (jwolski) [inline test/swim_test.js]: eh?

Comment (mrhooray) [inline test/swim_test.js]: mocking out setTimeout used by suspicion not only does not return a timeout object, also behaves differently from code running in prod

by passing suspicionTimeout in test, suspicion would behave close to prod and this test still runs deterministically

**Comment by jwolski** [comment]
looking good. you'll need a rebase now with changes. i want to spend some more time on that timeout object bug you're referencing too.

**Pull request merged**